src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -130,29 +135,32 @@ public static unsafe ulong BigMul(ulong a, ulong b, out ulong low)
 
             static ulong SoftwareFallback(ulong a, ulong b, out ulong low)
             {
-                // It's adoption of algorithm for multiplication
+                // Adaptation of algorithm for multiplication
                 // of 32-bit unsigned integers described
                 // in Hacker's Delight by Henry S. Warren, Jr. (ISBN 0-201-91465-4), Chapter 8
                 // Basically, it's an optimized version of FOIL method applied to
                 // low and high dwords of each operand
-                const ulong lowBitsMask = 0xFFFFFFFFU;
 
-                ulong al = a & lowBitsMask;
-                ulong ah = a >> 32;
-                ulong bl = b & lowBitsMask;
-                ulong bh = b >> 32;
+                uint al = (uint)a;
```

Comment:
Changing the type of these to `uint` makes the JIT to use 32bitx32bit->64bit mul instruction instead of using much slower 64bitx64bit->64bit helper. 

This change makes the fallback a lot faster (1.5x+) on 32-bit platform and a bit slower on x64 (~5%). Since 64-bit platforms are likely to use intrinsic path, this should be a good trade-off (the alternative would be to have separate implementations for 32-bit vs. 64-bit).

Reply:
Worth a comment maybe?